filebeat/input/file/state.go

```diff
@@ -18,82 +18,93 @@
+	Finished       bool              `json:"-"` // harvester state
+	Fileinfo       os.FileInfo       `json:"-"` // the file info
+	Source         string            `json:"source"`
+	Offset         int64             `json:"offset"`
+	Timestamp      time.Time         `json:"timestamp"`
+	TTL            time.Duration     `json:"ttl"`
+	Type           string            `json:"type"`
+	Meta           map[string]string `json:"meta"`
+	Data           interface{}       `json:"data,omitempty"` // arbitrary data of a comparator
+	ComparatorName string            `json:"comparator_name"`
+	FileStateOS    file.StateOS
+
+	identifier StateIdentifier
 }
 
 // NewState creates a new file state
-func NewState(fileInfo os.FileInfo, path string, t string, meta map[string]string) State {
+func NewState(fileInfo os.FileInfo, path string, t string, meta map[string]string, comparator IdentityComparator) State {
 	if len(meta) == 0 {
 		meta = nil
 	}
+
+	identifier := comparator.StateIdentifier()
+
 	return State{
-		Fileinfo:    fileInfo,
-		Source:      path,
-		Finished:    false,
-		FileStateOS: file.GetOSState(fileInfo),
-		Timestamp:   time.Now(),
-		TTL:         -1, // By default, state does have an infinite ttl
-		Type:        t,
-		Meta:        meta,
+		Fileinfo:       fileInfo,
+		Source:         path,
+		Finished:       false,
+		FileStateOS:    file.GetOSState(fileInfo),
+		Timestamp:      time.Now(),
+		TTL:            -1, // By default, state does have an infinite ttl
+		Type:           t,
+		Meta:           meta,
+		ComparatorName: comparator.Name(),
+		Data:           identifier.DataToSerialize(),
+		identifier:     identifier,
 	}
 }
 
-// ID returns a unique id for the state as a string
+// ID returns the ID of the state.
 func (s *State) ID() string {
-	// Generate id on first request. This is needed as id is not set when converting back from json
-	if s.Id == "" {
-		if len(s.Meta) == 0 {
-			s.Id = s.FileStateOS.String()
-		} else {
-			hashValue, _ := hashstructure.Hash(s.Meta, nil)
-			var hashBuf [17]byte
-			hash := strconv.AppendUint(hashBuf[:0], hashValue, 16)
-			hash = append(hash, '-')
-
-			fileID := s.FileStateOS.String()
+	return s.identifier.ID(s)
+}
 
-			var b strings.Builder
-			b.Grow(len(hash) + len(fileID))
-			b.Write(hash)
-			b.WriteString(fileID)
+// IsEqual checks if the two states point to the same file.
+func (s *State) IsEqual(c *State) bool {
+	return s.ID() == c.ID()
+}
 
-			s.Id = b.String()
-		}
+// SetupIdentifier initializes an identifier based on the comparator name.
+func (s *State) SetupIdentifier() error {
+	cfg, err := common.NewConfigFrom(map[string]interface{}{s.ComparatorName: nil})
+	if err != nil {
+		return fmt.Errorf("cannot init configuration from comparator name: %+v", err)
+	}
+	cmp, err := NewIdentityComparator(cfg)
+	if err != nil {
+		return fmt.Errorf("cannot create identity comparator: %+v", err)
 	}
 
-	return s.Id
+	s.identifier = cmp.StateIdentifier()
```

Comment:
If we make the ID 'eager' and serialize it as well, would this remove the need for `SetupIdentifier`?